ENH: graph Laplacian as LinearOperator, add dtype and symmetrized options

lobpcg opened the pull request:
#### Reference issue
Closes #10826, closes #10827, closes #10006, closes #15420 and closes #15419

Allows to rejuvenate https://github.com/scikit-learn/scikit-learn/pull/14636 which would then close  https://github.com/scikit-learn/scikit-learn/issues/8834 and https://github.com/scikit-learn/scikit-learn/pull/8842.

#### What does this implement/fix?
https://docs.scipy.org/doc/scipy-1.3.0/reference/generated/scipy.sparse.csgraph.laplacian.html currently only supports calculation the Graph Laplacian as a matrix. 

The spectral embedding function computes the laplacian matrix, which is the same size as the affinity matrix. This temporarily doubles the memory usage in many cases. 

However, sparse linear and eigenvalue solvers in scipy also support matrix input in the form of a LinearOperator, i.e., matrix-free.

Describe the solution you'd like
Introduce support for LinearOperator (and lambda-function) output, so that no new matrices are created, and no existing matrixes are modified, providing the ultimate memory-efficient solution. 

A free bonus: introducing the new dtype and symmetrized options which are especially helpful for  LinearOperator setup

lobpcg commented:
The PR turned out to be MUCH more effort, coding, and time consuming that I have originally anticipated. Could someone please review?  It is fully completed and passed all the old and new unit tests. I could add more examples to the docstring, if asked, e.g. from https://en.wikipedia.org/wiki/Laplacian_matrix

lobpcg commented:
Could someone please review? It is fully completed and passed all the old and new unit tests. The checks failed are all unrelated to this PR. Ready to merge, just waiting for reviews.

lobpcg commented:
Could someone please review? It is fully completed and passed all the old and new unit tests. The checks failed are all unrelated to this PR. Ready to merge, just waiting for reviews.

lobpcg commented:
Could someone please review? It is fully completed and passed all the old and new unit tests. The checks failed are all unrelated to this PR. Ready to merge, just waiting for reviews.

It's nearly a month old already. I start forgetting what this PR is about.

mdhaber commented:
I can't review this, but replying so that you know someone is paying attention. 

Note that delays in reviews happen frequently. (e.g. my gh-14576 waited from August to February). Keep in mind that pretty much everyone is a volunteer right now and there are [300+ other open PRs](https://github.com/scipy/scipy/pulls).

I know you've contributed before, but remember from [Ways to Contribute](http://scipy.github.io/devdocs/dev/hacking.html#hacking):
> ... the first step is to discuss on the scipy-dev mailing list. All new features, as well as changes to existing code, are discussed and decided on there. You can, and probably should, already start this discussion before your code is finished. Remember that in order to be added to SciPy your code will need to be reviewed by someone else, so try to find someone willing to review your work while you’re at it.

I didn't see an email to the [scipy-dev](https://mail.python.org/mailman3/lists/scipy-dev.python.org/) list, so I'd suggest sending one and explaining why this is important. In the future, consider doing that before developing the feature to make sure others agree, and arrange for a reviewer at that point so you don't need to wait so long after writing the code.

lobpcg commented:
> I didn't see an email to the [scipy-dev](https://mail.python.org/mailman3/lists/scipy-dev.python.org/) list, so I'd suggest sending one and explaining why this is important. In the future, consider doing that before developing the feature to make sure others agree, and arrange for a reviewer at that point so you don't need to wait so long after writing the code.
> 
> I'm sorry it's frustrating

@mdhaber  It's not frustrating. I am just trying to attract someone's attention, literally while I remember - more time passes, more difficult for me would be to react to suggestions to modify my own code that I have written already a month ago. This PR is a rather significant rewrite, although fortunately limited to only 2 files. 

Following your advice, I've attempted to create a new thread in SciPy-Dev, but do not see it appearing posted. I guess the list is moderated?

steppi commented:
Hi @lobpcg, this looks really cool. I think I'd be qualified to review this, but my time is very limited now. It will probably be around 5 months before I have enough free time to delve in. If no one else is available to review by then, I'd be happy to do it. I hope you'll be able to find a reviewer sooner than that but that it at least helps to have an upper bound.

perimosocordiae commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -83,78 +121,222 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+
+def _linearoperator(mv, shape, dtype):
+    mv_as_lo = LinearOperator(matvec=mv, matmat=mv, shape=shape, dtype=dtype)
+    return mv_as_lo
```
my preference would be to `return LinearOperator(matvec=mv, matmat=mv, shape=shape, dtype=dtype)` without the local variable assignment

perimosocordiae commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -83,78 +121,222 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+    if dtype is None:
+        dtype = graph.dtype
+
+    if form != "array":
```
As these are effectively different functions, I think it would be easier to read if you split these into _laplacian_sparse_matrix() for the form='array' case and _laplacian_sparse_function() for the other cases. It looks like the only duplicated code would be the default value for `dtype`.

lobpcg replied:
Done. I kept the names _laplacian_sparse and _laplacian_dense for the original functions for the array output and introduced _laplacian_sparse_flo and _laplacian_dense_flo. Please have a look.

There are now some unused variables in function calls, which I need to figure out how to handle

perimosocordiae commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -83,78 +121,222 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
-        m /= w[:, np.newaxis]
-        m *= -1
-        _setdiag_dense(m, 1 - isolated_node_mask)
+    if dtype is None:
```
`dtype` should never be None here, as you already assigned it to `graph.dtype` a few lines above.

lobpcg replied:
Well, in some inherited unit tests, the graph is a list with no dtype, so graph.dtype is None. Then `array` or `asarray` turns the graph into the array m, where by default, the data-type of m is inferred from the input data - the graph. 
Removing the alleged duplication as you propose makes these unit tests fail, as I have observed.

If graph.dtype is not None, it should be the same as m.dtype, indeed.

perimosocordiae replied:
I'm surprised that the `graph.dtype` assignment above would work in the case where `graph` is a list. Can we do the `if copy:` part first, and then remove the initial `dtype` assignment?

lobpcg replied:
Thanks, done. That would rely on array or asarray to infer dtype from the graph.

lobpcg replied:
done

perimosocordiae commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -83,78 +121,222 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+    if dtype is None:
+        dtype = m.dtype
+
+    if form != "array":
```
Same comment here about splitting this into two separate functions

lobpcg replied:
done

perimosocordiae commented on scipy/sparse/csgraph/tests/test_graph_laplacian.py:
```diff
@@ -99,86 +100,255 @@ def _assert_allclose_sparse(a, b, **kwargs):
+                np.issubdtype(mat.dtype, np.signedinteger)
+                or np.issubdtype(mat.dtype, np.uint)
+            )
+        ):
```
Maybe this is a sign that we should pull out this expression into a little helper function, because the black formatting is really awkward here (and elsewhere in this file).

lobpcg replied:
Done

perimosocordiae reviewed (changes requested):
I did a quick review pass for style / readability, without thinking too much about the actual math involved here. I'll try to give this a deeper look again soon.

lobpcg commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -37,31 +41,58 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+    form: `array`, or `function`, or `lo`
+        Determins the format of the output.
+        E.g., `lo` results in the format of `LinearOperator`.
+        Choosing `function`, or `lo` always avoids doubling
```
Done, thanks! Please check https://github.com/scipy/scipy/pull/15391/commits/c0e26e3bcd9e9599d37ec041366db773d5ed1aa4

lobpcg replied:
Done, thanks! Please check https://github.com/scipy/scipy/pull/15391/commits/c0e26e3bcd9e9599d37ec041366db773d5ed1aa4

lobpcg commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -83,78 +121,222 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
-        m.data /= w[m.col]
-        m.data *= -1
-        m.setdiag(1 - isolated_node_mask)
+def _md(m, d):
```
replaced with `laplace`

lobpcg commented:
> I did a quick review pass for style / readability, without thinking too much about the actual math involved here. I'll try to give this a deeper look again soon.

@perimosocordiae Thanks for your prompt review! I have addressed all your suggestions. I have taken the liberty to resolve conversations about trivial changes made, and left open conversations where you may want to have another look.

perimosocordiae commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -83,72 +124,249 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+            return md, w.astype(dtype, copy=False)
+        elif form == "lo":
+            m = _linearoperator(md, shape=graph.shape, dtype=dtype)
+            return m, w.astype(dtype, copy=False)
```
Looks like we're missing the final `else:` branch here.

```
else:
  raise ValueError(f"Invalid form: {repr(form)}")
```

lobpcg replied:
done

lobpcg commented:
@perimosocordiae Thanks for your prompt comments! I have made all the changes. I have taken the liberty to resolve conversations about trivial changes made, and left open conversations where you may want to have another look. Please close those yourself if you feel they are addressed to your satisfaction.

lobpcg commented:
I have greatly extended the documentation, added a couple of new tests, and made a few fixes for the new tests to pass. Cannot think of anything more to improve without additional external input. LGTM. 

The tests failed appear to be unrelated to this PR.

lobpcg commented:
I have greatly extended the documentation, added a couple of new tests, and made a few fixes for the new tests to pass. Cannot think of anything more to improve without additional external input. LGTM.

The tests failed appear to be unrelated to this PR.

mdhaber commented:
Yeah as @tupui mentioned, the Main refguide_asv_check failure appears to be real:

```

Failed example:
        L = csgraph.laplacian(G, symmetrized=True, form="lo")
Exception raised:
    Traceback (most recent call last):
      File "/opt/hostedtoolcache/Python/3.9.10/x64/lib/python3.9/doctest.py", line 1334, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest laplacian[46]>", line 1
        L = csgraph.laplacian(G, symmetrized=True, form="lo")
    IndentationError: unexpected indent


File "build/testenv/lib/python3.9/site-packages/scipy/sparse/csgraph/_laplacian.py", line 315, in laplacian
Failed example:
        _, eves = sparse.linalg.lobpcg(L, X, Y=Y, largest=False, tol=1e-3)
Exception raised:
    Traceback (most recent call last):
      File "/opt/hostedtoolcache/Python/3.9.10/x64/lib/python3.9/doctest.py", line 1334, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest laplacian[47]>", line 1
        _, eves = sparse.linalg.lobpcg(L, X, Y=Y, largest=False, tol=1e-3)
    IndentationError: unexpected indent


File "build/testenv/lib/python3.9/site-packages/scipy/sparse/csgraph/_laplacian.py", line 316, in laplacian
Failed example:
        print(cut + "-cut labels:\n", 1 * (eves[:, 0]>0))
Exception raised:
    Traceback (most recent call last):
      File "/opt/hostedtoolcache/Python/3.9.10/x64/lib/python3.9/doctest.py", line 1334, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest laplacian[48]>", line 1
        print(cut + "-cut labels:\n", 1 * (eves[:, 0]>0))
    IndentationError: unexpected indent
```

Fixing that will clear up scipy.scipy and probably `build_docs` and `build_docs_artifact`. You might fix `run_benchmarks` by merging main?

lobpcg commented:
@mdhaber thanks for checking! Maybe you could also please hint how to fix the indentation here? I just do not know how it should look like in docstrings

```
    >>> for cut in ["max", "min"]:
    >>>     G = -G
```

mdhaber commented:
@lobpcg continuation line gets `...` instead of `>>>`:
```
    >>> for cut in ["max", "min"]:
    ...     G = -G
```

lobpcg commented:
@mdhaber thnks, this issue is now fixed! 

Build docs fail due to
make: *** No rule to make target 'latex'.  Stop.
but makes html docs before that which look good. 
Other failing tests are time out - nothing I can do I guess.

mdhaber commented:
I will merge master for you. I think that will fix it.

tupui commented:
But I am not the merge was correct. Looking at the commands that got executed on CircleCI, it does not corresponds to what's in main.

mdhaber commented:
No it wasn't. I'm working on it.

mdhaber commented:
@lobpcg I think when you merged main it was out of date. Remember to `git fetch upstream` before `git merge upstream/main`, or remember to `git pull` on your local `main` if you do `git merge main`.

The other thing was 
![image](https://user-images.githubusercontent.com/6570539/159187523-13997651-02c7-46a2-9ff1-3da72e06d8ed.png)

Unless you're intending to modify a submodule in a PR, you don't want to commit these changes. If you ever see a submodule has changes when you do `git status`, do `git submodule update --init`. If that doesn't clear them, let us know; there were some additional steps that were required when we renamed `master`->`main`.

lobpcg commented:
@mdhaber thanks very much for your help! Not everyone is technically knowledgeable enough, e.g., I simply use GitHub Desktop, and just clicked on a button "update from main" there.... 

This appears to be a common issue with having PRs idle for too long without merging. Some contributors may just give up maintaining their PRs if it takes too much time/effort to keep them alive waiting for review/merge. One such sad example is https://github.com/scipy/scipy/pull/11829 which I would like to see finished. 

Anyway, one PR at the time for me - thanks again! I hope that this PR gets reviewed and merged before one would need again to save it from another cardiac arrest due to aging.

lobpcg commented:
GitHub Desktop unintended action.. Will try to reverse.

mdhaber commented:
> Not everyone is technically knowledgeable enough

Certainly. NP. I only learned `git` through years of working with SciPy.

> This appears to be a common issue with having PRs idle for too long without merging.

It's unfortunate, but that's the way it goes sometimes : / Does gh-11829 look good to you? A thorough review and approval from a non-maintainer helps, too. Whe you review a PR, are you able to click an "approve" radio button?

![image](https://user-images.githubusercontent.com/6570539/159191527-f3e13705-1261-48de-9b9e-2c0cee18ab91.png)

mdhaber commented:
Yeah this is exactly what we don't want. If you can't get GitHub Desktop to cooperate (or if you have trouble following the command line instructions I gave), LMK and I'll revert.

lobpcg commented:
It's failing, of course. But if you look at the files, it appears that my 2 files were also apparently affected by your changes? So you could not just revert because that would also remove my changes?

lobpcg commented:
I try reverting myself

lobpcg commented:
>  Does [gh-11829](https://github.com/scipy/scipy/pull/11829) look good to you? 

Mathematically, yes.  I also know and trust the author. 

Code-wise, probably - I am not competent enough to judge.

> A thorough review and approval from a non-maintainer helps, too. Whe you review a PR, are you able to click an "approve" radio button

Yes.

mdhaber commented:
Yay! The remaining error is unrelated to the PR.

tupui commented:
@rossbar is this something you could look into?

perimosocordiae commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -37,118 +41,509 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+
+
+def _laplacian_sparse_flo(graph, normed, axis, copy, form, dtype, symmetrized):
+    # The keyword argument ``copy`` is unused and has no effect here.
```
In this case I would write:

```python
del copy  # Unused argument
```

This ensures that the variable will actually not be used in the body, while also making the intent clear to readers.

tupui commented on scipy/sparse/csgraph/tests/test_graph_laplacian.py:
```diff
@@ -99,86 +100,244 @@ def _assert_allclose_sparse(a, b, **kwargs):
 
     if not copy:
-        if not (normed and (np.issubdtype(mat.dtype, np.signedinteger)
-                            or np.issubdtype(mat.dtype, np.uint))):
+        if not (normed and check_int_type(mat)):
             if type(mat) is np.ndarray:
                 assert_allclose(L, mat)
-            elif mat.format == 'coo':
+            elif mat.format == "coo":
                 _assert_allclose_sparse(L, mat)
 
 
-REAL_DTYPES = {np.intc, np.int_, np.longlong,
-               np.single, np.double, np.longdouble}
+INT_DTYPES = {np.intc, np.int_, np.longlong}
+REAL_DTYPES = {np.single, np.double, np.longdouble}
 COMPLEX_DTYPES = {np.csingle, np.cdouble, np.clongdouble}
 # use sorted tuple to ensure fixed order of tests
-DTYPES = tuple(sorted(REAL_DTYPES ^ COMPLEX_DTYPES, key=str))
+DTYPES = tuple(sorted(INT_DTYPES ^ REAL_DTYPES ^ COMPLEX_DTYPES, key=str))
 
 
 @pytest.mark.parametrize("dtype", DTYPES)
-@pytest.mark.parametrize("arr_type", [np.array,
-                                      sparse.csr_matrix,
-                                      sparse.coo_matrix])
+@pytest.mark.parametrize(
+    "arr_type", [np.array, sparse.csr_matrix, sparse.coo_matrix]
+)
 @pytest.mark.parametrize("copy", [True, False])
 @pytest.mark.parametrize("normed", [True, False])
 @pytest.mark.parametrize("use_out_degree", [True, False])
-def test_asymmetric_laplacian(use_out_degree, normed,
-                              copy, dtype, arr_type):
+def test_asymmetric_laplacian(use_out_degree, normed, copy, dtype, arr_type):
     # adjacency matrix
-    A = [[0, 1, 0],
-         [4, 2, 0],
-         [0, 0, 0]]
+    A = [[0, 1, 0], [4, 2, 0], [0, 0, 0]]
```
All style changes are not necessary which maybe account for half of the diff in the PR. We tend to not reformat whole files to preserve a useful history per line (and help reviews). It's usually fine to update the style on the function you are working on, but a lot of changes in this PR are not restricted to the proposed changes. I know you are probably just using black or another formatter automatically, but we cannot do this with SciPy.

I don't want to block this PR for that, given there is a maintainer approval, but please consider this next time.

tupui replied:
To be clear, we don't have a specific "tradition" 😅 But because we decided not to use Black for now, we cannot have people refactor whole files with Black as we go because it breaks history. New code can follow Black's formatting but only because compliant with pep8 and our linters.

If we merge your PR right now, and someone wants to know why one particular line is written like that, they will see your PR first. Then they will have to go trough the diff and lose some time to understand that the information they are looking for is not there and go to the next commit. SciPy is 20 years old, and styling changed a few time in this timeframe. Imagine if we reformat things with all style updates the community went through, we would be burying the real nature of the change deeper and deeper as we go.

With Black and sklearn, they made the decision at the package level and enforce it. Meaning they can mark specific commits to be style changes/Black only and hence not have this issue.

lobpcg replied:
I understand. 

I have black autoformat in my VS code on every save... I'll try to turn in off.

tupui replied:
Either this or only stage the lines you want to push to SciPy. Since you are using VSCode, you can manually select which lines to stage.

lobpcg replied:
@mdhaber could you please help with this? I have no taste for good/bad formatting and also would have difficulty separate necessary vs unnecessary formatting changes. To my recollection, some original text would not pass lint so formating was unavoidable. 

Everything else seems to work fine: all tests pass, and I have updated the docstring format (my other weakness) as proposed by @tupui .

tupui replied:
@mdhaber if you are up for, updating would be better. I could have a look tomorrow too. With a visual git it's fairly easy, I just go over the changes and reverts the lies in chunks with one click.

tupui replied:
Sure no problem 😉

tupui replied:
Done. The diff is more reasonable to review.

tupui commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -37,118 +41,510 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+    The constant vector of ones is always a trivial eigenvector
+    of the non-normalized Laplacian to be filtered out:
+
+    >>> Y=np.ones((N, 1))
```
```suggestion
    >>> Y = np.ones((N, 1))
```

tupui commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -37,118 +41,510 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+
+    Fix a random seed ``rng`` and add a random sparse noise to the graph ``G``:
+
+    >>> rng = np.random.default_rng(0)
```
The seed will be added automatically in the doc. We go to great extent to prevent using any fixed seed values as people tend to just copy past them.

Locally you can use `1638083107694713882823079058616272161` to make the change on the values the code should return. (This will get into the dev doc soon, we know it's not clear for now.)

```suggestion
    >>> rng = np.random.default_rng()
```

tupui commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -37,118 +41,510 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+
+def _laplace_normed(m, d, nd):
+    laplace = _laplace(m, d)
+    return (
+        lambda v: nd[:, np.newaxis] * laplace(v * nd[:, np.newaxis])
+           )
```
Never seen such formatting
```suggestion
    return (
        lambda v: nd[:, np.newaxis] * laplace(v * nd[:, np.newaxis])
    )
```

tupui commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -37,118 +41,510 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+
+def _laplace_normed_sym(m, d, nd):
+    laplace_sym = _laplace_sym(m, d)
+    return (
+        lambda v: nd[:, np.newaxis] * laplace_sym(v * nd[:, np.newaxis])
+           )
```
```suggestion
    return (
        lambda v: nd[:, np.newaxis] * laplace_sym(v * nd[:, np.newaxis])
    )
```

tupui commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -37,118 +41,510 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+def _laplacian_sparse(graph, normed, axis, copy, form, dtype, symmetrized):
+
+    if form != "array":
+        raise ValueError('{form} must be "array"'.format(form=repr(form)))
```
```suggestion
        raise ValueError(f'{repr(form)} must be "array"')
```

tupui commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -37,118 +41,510 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+def _laplacian_dense(graph, normed, axis, copy, form, dtype, symmetrized):
+
+    if form != "array":
+        raise ValueError('{form} must be "array"'.format(form=repr(form)))
```
```suggestion
        raise ValueError('{repr(form)} must be "array"')
```

tupui commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -37,118 +41,510 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+            m = _linearoperator(md, shape=graph.shape, dtype=dtype)
+            return m, w.astype(dtype, copy=False)
+        else:
+            raise ValueError(f"Invalid form: {repr(form)}")
```
The new validation errors need to be tested using `pytest.raises(ValueError, match=...)`

lobpcg replied:
@tupui Could you please help and propose the specific change here? I'll then follow your sample in other occurrences.

tupui replied:
What I mean is that in the tests you need to call the function in a way that would make this warning getting raised. In the tests you would have:
```python
with pytest.raises(ValueError, match="Invalid form: 'toto'"):
    laplacian(..., form='toto')
```

By the way, noticing this now. Instead of `f"{repr(form)}"` what you want to do is `f"{form!r}"`.

lobpcg replied:
But the line you comment on is not in the test?

tupui replied:
It's a new message no? If I am not mistaking I did not see any test using a wrong `form` and checking the correct message is raised.

lobpcg replied:
got it, will do, thanks!

tupui replied:
Thanks 😃

tupui reviewed (changes requested):
Thanks @lobpcg. I did a review of the code style, trusting @perimosocordiae's review for the logic/maths so I am not going to go over it.

tupui commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -37,118 +41,508 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+def _laplacian_sparse(graph, normed, axis, copy, form, dtype, symmetrized):
 
+    if form != "array":
+        raise ValueError(f'{form!r} must be "array"')
```
This one as well is missing a test

lobpcg replied:
This is an internal function, not exposed to the test. "form" is actually unused here, so we can replace this with 
del form
if you like? It does check the internal logic of the parent function, however, so I would keep it.

tupui replied:
If this path is never going to be used, it should not be there. You could use `del form` as you did above.

lobpcg replied:
Could you please, while editing it?

lobpcg commented:
@tupui @mdhaber - thanks for your help! Anything else I need to do?

tupui reviewed (approved):
Thank you @lobpcg, LGTM 😃 Since there is an approval from @perimosocordiae, I will give it until Friday and merge if there is no more comments. (failure is unrelated AFAICT and previous run was green)

Clicked too early. refguide failure is real.

tupui commented on scipy/sparse/csgraph/_laplacian.py:
```diff
@@ -37,88 +41,416 @@ def laplacian(csgraph, normed=False, return_diag=False, use_out_degree=False,
+    max-cut labels:
+    [0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0 1 0]
+    min-cut labels:
+    [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1]
```
```suggestion
    [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
```

tupui commented:
> @tupui I somehow no longer see the outcome of the tests and thus the generated documentation. I assume that you have checked that the latest commit outcome was successful.

GitHub is currently having an outage. We have to wait I am afraid.

lobpcg commented:
@tupui On that last change: zeros and ones here are labels, so if they swap the result is still correct. I hope that the fixed seed makes the output deterministic, but would not be too surprised if zeros and ones swap if one runs this on different systems. Could this present a problem for  refguide ?

tupui commented:
> @tupui On that last change: zeros and ones here are labels, so if they swap the result is still correct. I hope that the fixed seed makes the output deterministic, but would not be too surprised if zeros and ones swap if one runs this on different systems. Could this present a problem for refguide ?

The output is deterministic for the arrays yes. If `lobpcg` and the other functions you are calling are deterministic, then there should be no problems.

lobpcg commented:
> 

Does refguide require deterministic output matching the docstring text? 

In this example, I feel that the output may be not deterministic when run on different systems, compiled with different flags etc. If this presents a problem, I can fix that with another small edit

lobpcg commented:
@tupui LGTM including the docs

tupui commented:
No complains, it's Friday, merging. Thanks again @lobpcg

tupui commented:
@lobpcg can you add a release note in https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.9.0?

tupui commented:
Thank you @lobpcg 😃
